File: Aug/Day25/diagonal-traverse.py
#  Time: O(M * N)  ::  Space: O(min(N, M))
class Solution:
    def findDiagonalOrder(self, matrix: List[List[int]]) -> List[int]:
        
        if not matrix or not matrix[0]:
            return []

        n, m = len(matrix), len(matrix[0])
        
        result, temp = [], []
        
       
        for headIndex in range(n+m-1):            
            temp.clear()
            row, col = 0 if headIndex < m else headIndex - m + 1, headIndex if headIndex < m else m - 1

            while row < n and col >= 0:
                temp.append(matrix[row][col])
                row += 1
                col -= 1
            if headIndex % 2 == 0:
                result.extend(temp[::-1])
            else:
                result.extend(temp)
        return result        

File: Aug/Day25/diagonal-traverse.java
//  Time: O(M * N)  ::  Space: O(min(N, M))
class Solution {
    public int[] findDiagonalOrder(int[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) 
            return new int[0];

        int n = matrix.length, m = matrix[0].length;
        List<Integer> result = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();

        for (int headIndex = 0; headIndex < n + m - 1; headIndex++) {
            temp.clear();
            int row = (headIndex < m) ? 0 : headIndex - m + 1;
            int col = (headIndex < m) ? headIndex : m - 1;

            while (row < n && col >= 0) {
                temp.add(matrix[row][col]);
                row++;
                col--;
            }

            if (headIndex % 2 == 0)
                Collections.reverse(temp);

            result.addAll(temp);
        }

        return result.stream().mapToInt(i -> i).toArray();
    }
}
